google-cloud-spanner/lib/google/cloud/spanner/commit.rb

```diff
@@ -156,12 +165,19 @@ def insert table, *rows
           rows = Array(rows).flatten
           return rows if rows.empty?
           rows.compact
-          rows.delete_if(&:empty?)
+          rows.delete_if { |row| row.respond_to?(:empty?) && row.empty? }
```

Comment:
This method looks nearly identical to `upsert` (above), everything besides the Mutation field being used. And the complexity of the logic is starting to get to the point where maybe it makes sense to dedup the methods so they don't get out of sync.

Edit: and same with `update` below.

Edit 2: and `replace`.

Reply:
Done, I deduped the main part. PLMK if you have any thoughts on naming though!